File: python_scripts/03_qc.py
#!/usr/bin/env python3
"""
Step 3: Quality Control for single-cell RNA-seq data.

- Annotate QC metrics (gene counts, mitochondrial percentage).
- Generate violin and scatter plots grouped by metadata.
- Filter cells and genes based on quality thresholds.
- Save filtered data and generate an HTML QC report.
"""

import sys
import yaml
import datetime
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import scanpy as sc
from logzero import logger
from pipeline_utils import setup_dirs_logs

warnings.simplefilter(action="ignore", category=FutureWarning)


# -----------------------------
# Load Config and Parameters
# -----------------------------
if len(sys.argv) < 3:
    print("Usage: step3_qc.py <project_name> <config.yaml>")
    sys.exit(1)

PROJ_NAME = sys.argv[1]
CONFIG_FILE = Path(sys.argv[2])

if not CONFIG_FILE.exists():
    print(f"Config file not found: {CONFIG_FILE}")
    sys.exit(1)

with open(CONFIG_FILE, "r") as f:
    cfg = yaml.safe_load(f)

# Enforce strict QC parameters
try:
    QC_PARAMS = cfg["qc"]
    MIN_GENES = QC_PARAMS["min_genes"]
    MAX_PCT_MT = QC_PARAMS["max_pct_mt"]
    MIN_CELLS_PER_GENE = QC_PARAMS["min_cells_per_gene"]
except KeyError as e:
    print(f"Missing QC parameter in config.yaml: {e}")
    sys.exit(1)

RESULTS_DIR, FIGURE_DIR, LOG_FILE = setup_dirs_logs("03_log.txt", PROJ_NAME)
INPUT_FILE = RESULTS_DIR / "02_merged_data.h5ad"
OUTPUT_FILE = RESULTS_DIR / "03_filtered_data.h5ad"
HTML_REPORT = RESULTS_DIR / "03_report.html"

sc.settings.verbosity = 1  # minimal console output
sc.settings.logfile = LOG_FILE  # direct Scanpy logs to log file
sc.settings.figdir = FIGURE_DIR  # save plots automatically here

logger.info("Step 3 started: Quality Control")


# -----------------------------
# Helper Functions
# -----------------------------
def load_data(input_file: Path):
    """Load AnnData object from previous step."""
    if not input_file.exists():
        logger.error(f"Input file not found: {input_file}")
        sys.exit(1)
    adata = sc.read(input_file)
    logger.info(f"Loaded data: {adata.n_obs} cells × {adata.n_vars} genes")
    return adata


def determine_grouping(adata):
    """Choose grouping column for QC plots."""
    if "treatment" in adata.obs.columns:
        return ["treatment"]
    elif "sample" in adata.obs.columns:
        return ["sample"]
    logger.error("No 'treatment' or 'sample' column found in adata.obs.")
    sys.exit(1)


def annotate_qc(adata):
    """Add mitochondrial gene flag and compute QC metrics."""
    adata.var["mt"] = adata.var_names.str.upper().str.startswith("MT-")
    sc.pp.calculate_qc_metrics(adata, qc_vars=["mt"], inplace=True)


def save_qc_plots(adata, group_cols, tag):
    """Save QC violin and scatter plots for each grouping column."""
    for group in group_cols:
        sc.pl.violin(
            adata,
            ["n_genes_by_counts", "total_counts", "pct_counts_mt"],
            groupby=group,
            jitter=0.4,
            rotation=45,
            save=f"_qc_violin_by_{group}_{tag}.png",
            show=False,
        )
        sc.pl.scatter(
            adata,
            x="total_counts",
            y="pct_counts_mt",
            color=group,
            save=f"_qc_scatter_mt_by_{group}_{tag}.png",
            show=False,
        )
        sc.pl.scatter(
            adata,
            x="total_counts",
            y="n_genes_by_counts",
            color=group,
            save=f"_qc_scatter_genes_by_{group}_{tag}.png",
            show=False,
        )
        logger.info(f"Saved QC plots grouped by {group} ({tag} filtering)")


def filter_data(adata):
    """Apply cell/gene QC filters and return filtered AnnData."""
    pre_shape = adata.shape
    sc.pp.filter_cells(adata, min_genes=MIN_GENES)
    sc.pp.filter_genes(adata, min_cells=MIN_CELLS_PER_GENE)
    adata = adata[adata.obs["pct_counts_mt"] < MAX_PCT_MT, :]
    post_shape = adata.shape

    logger.info(f"Filtering complete — Before: {pre_shape}, After: {post_shape}")
    logger.info(
        f"Applied thresholds — min_genes={MIN_GENES}, "
        f"min_cells_per_gene={MIN_CELLS_PER_GENE}, max_pct_mito={MAX_PCT_MT}%"
    )
    return adata, pre_shape, post_shape


def save_html_report(html_file, figure_dir, pre_shape, post_shape, group_cols):
    """Generate a static HTML report embedding QC plots."""
    figure_dir = figure_dir.resolve()
    now = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    with open(html_file, "w") as f:
        f.write(f"<html><head><title>QC Report</title></head><body>\n")
        f.write("<h1>Single-Cell QC Report</h1>\n")
        f.write(f"<p><b>Date:</b> {now}</p>\n")
        f.write(
            f"<p><b>Initial shape:</b> {pre_shape[0]} cells × {pre_shape[1]} genes</p>\n"
        )
        f.write(
            f"<p><b>Post-filtering shape:</b> {post_shape[0]} cells × {post_shape[1]} genes</p>\n"
        )
        f.write(
            f"<p><b>QC thresholds:</b> min_genes={MIN_GENES}, "
            f"min_cells_per_gene={MIN_CELLS_PER_GENE}, "
            f"max_pct_mito={MAX_PCT_MT}%</p>\n"
        )

        f.write("<h2>QC Metrics Before Filtering</h2>\n")
        for group in group_cols:
            f.write(
                f'<img src="{figure_dir}/violin_qc_violin_by_{group}_before.png" width="600"><br>\n'
            )
            f.write(
                f'<img src="{figure_dir}/scatter_qc_scatter_mt_by_{group}_before.png" width="600"><br>\n'
            )
            f.write(
                f'<img src="{figure_dir}/scatter_qc_scatter_genes_by_{group}_before.png" width="600"><br>\n'
            )

        f.write("<h2>QC Metrics After Filtering</h2>\n")
        for group in group_cols:
            f.write(
                f'<img src="{figure_dir}/violin_qc_violin_by_{group}_after.png" width="600"><br>\n'
            )
            f.write(
                f'<img src="{figure_dir}/scatter_qc_scatter_mt_by_{group}_after.png" width="600"><br>\n'
            )
            f.write(
                f'<img src="{figure_dir}/scatter_qc_scatter_genes_by_{group}_after.png" width="600"><br>\n'
            )

        f.write("</body></html>\n")
    logger.info(f"HTML QC report generated at: {html_file}")


# -----------------------------
# Main Pipeline
# -----------------------------
def main():
    adata = load_data(INPUT_FILE)
    group_cols = determine_grouping(adata)

    annotate_qc(adata)
    save_qc_plots(adata, group_cols, tag="before")

    adata_filtered, pre_shape, post_shape = filter_data(adata)
    save_qc_plots(adata_filtered, group_cols, tag="after")

    adata_filtered.write(OUTPUT_FILE)
    logger.info(f"Filtered data saved to {OUTPUT_FILE}")

    save_html_report(HTML_REPORT, FIGURE_DIR, pre_shape, post_shape, group_cols)
    logger.info("Step 3 complete: QC applied, filtered data saved.")


if __name__ == "__main__":
    main()

File: R_scripts/03_qc.R
# -----------------------------
# Step 3: Quality Control
# -----------------------------

# Clear environment, graphics, and console
rm(list = ls())            # Remove all variables from the workspace
graphics.off()             # Close all open graphics devices
cat("\014")                # Clear the console (works in RStudio)

library(Seurat)
library(ggplot2)
library(dplyr)

# -----------------------------
# Parameters and Paths
# -----------------------------
output_dir <- "results"
plot_dir <- "figures"
input_file <- file.path(output_dir, "02_merged_data.rds")
output_file <- file.path(output_dir, "03_filtered_data.rds")
report_file <- file.path(output_dir, "03_report.html")
log_file <- file.path(output_dir, "03_log.txt")

if (!dir.exists(output_dir)) dir.create(output_dir, recursive = TRUE)
if (!dir.exists(plot_dir)) dir.create(plot_dir, recursive = TRUE)

qc_params <- list(
  min_genes = 200,
  max_pct_mt = 10,
  min_cells_per_gene = 3
)


# -----------------------------
# Functions
# -----------------------------
# Function to set the log message
log_msg <- function(msg) {
  
  # Generate current timestamp in YYYY-MM-DD HH:MM:SS format
  timestamp <- format(Sys.time(), "%Y-%m-%d %H:%M:%S")
  
  # Construct log line with timestamp prefix
  line <- sprintf("[%s] %s", timestamp, msg)
  
  # Print log message to console
  cat(line, "\n")
  
  # Append log message to the specified log file
  # Assumes 'log_file' is a global variable set elsewhere in the script
  write(line, file = log_file, append = TRUE)
}

# Function to generate and save QC plots
save_qc_plots <- function(seurat_obj, group_label, suffix) {
  # Construct file paths for each plot 
  # (violin, scatter with percent.mt, scatter with nFeature_RNA)
  vln_file <- file.path(plot_dir, sprintf("qc_violin_by_%s_%s.png", group_label, suffix))
  scatter_mt_file <- file.path(plot_dir, sprintf("qc_scatter_mt_by_%s_%s.png", group_label, suffix))
  scatter_gene_file <- file.path(plot_dir, sprintf("qc_scatter_genes_by_%s_%s.png", group_label, suffix))
  
  # Generate violin plots of key QC metrics 
  # (number of features, counts, % mitochondrial content)
  g1 <- VlnPlot(
    seurat_obj, 
    features = c("nFeature_RNA", "nCount_RNA", "percent.mt"),
    group.by = group_label,    # Group cells by the specified metadata (e.g., treatment, sample)
    pt.size = 0.1,             # Make individual cell points small to reduce overplotting
    ncol = 3                   # Arrange violin plots in 3 columns
  )
  # Save the violin plot
  ggsave(vln_file, g1, width = 10, height = 4)  
  
  # Generate scatter plot: total counts vs. % mitochondrial reads
  g2 <- FeatureScatter(
    seurat_obj, 
    "nCount_RNA", 
    "percent.mt"
  )
  # Save the scatter plot
  ggsave(scatter_mt_file, g2, width = 10, height = 5)  
  
  # Generate scatter plot: total counts vs. number of detected genes
  g3 <- FeatureScatter(
    seurat_obj, 
    "nCount_RNA", 
    "nFeature_RNA"
  )
  # Save the scatter plot
  ggsave(scatter_gene_file, g3, width = 10, height = 5) 
  
  log_msg(sprintf("Saved QC plots (%s) for %s", suffix, group_label))
}


log_msg("Step 3 started: Quality Control")

# -----------------------------
# Load Data
# -----------------------------
seurat_obj <- readRDS(input_file)
log_msg(sprintf("Loaded merged data: %d cells × %d genes", ncol(seurat_obj), nrow(seurat_obj)))

# -----------------------------
# QC Metrics
# -----------------------------
# Calculate the percentage of reads mapping to mitochondrial genes (genes starting with "MT-")
# and add it as a metadata column named 'percent.mt'
seurat_obj[["percent.mt"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")

# List of metadata columns to group QC plots
group_cols <- c("treatment") 

# Generate and save QC plots for each group before filtering
for (group in group_cols) {
  save_qc_plots(seurat_obj, group, "before")
}

# -----------------------------
# Filtering
# -----------------------------
# Record the initial dimensions (genes × cells) before filtering
pre_shape <- dim(seurat_obj)

# Filter out low-quality cells:
# - Keep cells with at least min_genes expressed
# - Keep cells with mitochondrial percentage below the threshold
seurat_obj <- subset(
  seurat_obj,
  subset = nFeature_RNA >= qc_params$min_genes & percent.mt <= qc_params$max_pct_mt
)

# Filter out lowly expressed genes:
# - Count how many cells each gene is detected in (non-zero counts)
gene_counts <- Matrix::rowSums(GetAssayData(seurat_obj, assay = "RNA", layer = "counts") > 0)

# - Retain genes expressed in at least min_cells_per_gene cells
genes_to_keep <- names(gene_counts[gene_counts >= qc_params$min_cells_per_gene])

# - Subset Seurat object to keep only those genes
seurat_obj <- subset(seurat_obj, features = genes_to_keep)

# Record dimensions after filtering
post_shape <- dim(seurat_obj)

# Log the filtering summary (cells and genes before vs. after)
log_msg(sprintf(
  "Filtering complete — Before: %d cells × %d genes; After: %d cells × %d genes",
  pre_shape[2], pre_shape[1], post_shape[2], post_shape[1]
))

# Log applied QC thresholds for traceability
log_msg(sprintf(
  "Applied thresholds — min_genes: %d, min_cells_per_gene: %d, max_pct_mito: %d%%",
  qc_params$min_genes, qc_params$min_cells_per_gene, qc_params$max_pct_mt
))

# -----------------------------
# QC After Filtering
# -----------------------------
# Generate and save QC plots for each group after filtering
for (group in group_cols) {
  save_qc_plots(seurat_obj, group, "after")
}

# Save the filtered Seurat object to file for downstream analysis
saveRDS(seurat_obj, output_file)
log_msg(sprintf("Filtered data saved to %s", output_file))


# -----------------------------
# HTML Report
# -----------------------------
# Initialize HTML content as a character vector (each element = one line)
html <- c(
  "<html><head><title>QC Report</title></head><body>",
  "<h1>Single-Cell QC Report</h1>",
  # Insert the current date and time
  sprintf("<p><b>Date:</b> %s</p>", Sys.time()),
  # Insert initial data shape before filtering (cells × genes)
  sprintf("<p><b>Initial shape:</b> %d cells × %d genes</p>", pre_shape[2], pre_shape[1]),
  # Insert data shape after filtering
  sprintf("<p><b>Post-filtering shape:</b> %d cells × %d genes</p>", post_shape[2], post_shape[1]),
  # Insert QC thresholds used for filtering
  sprintf("<p><b>QC thresholds:</b> min_genes = %d, min_cells_per_gene = %d, max_pct_mito = %d%%</p>",
          qc_params$min_genes, qc_params$min_cells_per_gene, qc_params$max_pct_mt),
  "<h2>QC Metrics Before Filtering</h2>"
)

# Add plots for each group (e.g. "treatment") before filtering
for (group in group_cols) {
  html <- c(html,
            # Violin plot of QC metrics
            sprintf('<img src="../%s/qc_violin_by_%s_before.png" width="600"/><br>', plot_dir, group),
            # Scatter plot: nCount_RNA vs percent.mt
            sprintf('<img src="../%s/qc_scatter_mt_by_%s_before.png" width="600"/><br>', plot_dir, group),
            # Scatter plot: nCount_RNA vs nFeature_RNA
            sprintf('<img src="../%s/qc_scatter_genes_by_%s_before.png" width="600"/><br>', plot_dir, group))
}

# Add section header for after-filtering plots
html <- c(html, "<h2>QC Metrics After Filtering</h2>")

# Add plots for each group after filtering
for (group in group_cols) {
  html <- c(html,
            # Violin plot of QC metrics
            sprintf('<img src="../%s/qc_violin_by_%s_after.png" width="600"/><br>', plot_dir, group),
            # Scatter plot: nCount_RNA vs percent.mt
            sprintf('<img src="../%s/qc_scatter_mt_by_%s_after.png" width="600"/><br>', plot_dir, group),
            # Scatter plot: nCount_RNA vs nFeature_RNA
            sprintf('<img src="../%s/qc_scatter_genes_by_%s_after.png" width="600"/><br>', plot_dir, group))
}

# Close the HTML tags
html <- c(html, "</body></html>")

# Write the HTML content to file
writeLines(html, report_file)
log_msg(sprintf("HTML QC report generated at: %s", report_file))


cat("✅ Step 3 complete: Quality control applied, filtered data saved.\n")
